New functionality for astroquery.mpc

mommermi opened the pull request:
This PR adds a function to query the [Minor Planet Center observations database](https://minorplanetcenter.net/db_search). 

Docs and tests are included. `CHANGES` file has not yet been updated.

This PR also adds `astroquery.mpc` to `astroquery.solarsystem`.

pep8speaks commented:
Hello @mommermi! Thanks for submitting the PR.



 - In the file [`astroquery/mpc/core.py`](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py), following are the PEP8 issues :

> [Line 776:60](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L776): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 785:57](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L785): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 789:72](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L789): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 854:66](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L854): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 865:51](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L865): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 866:67](https://github.com/astropy/astroquery/blob/f82f870bb95373941120d66ff1121c2997adaf82/astroquery/mpc/core.py#L866): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

bsipocz commented on astroquery/mpc/tests/data/mpc_obs.dat:
```diff
@@ -0,0 +1 @@
+[{"designation":"1998 QS55","discovery_flag":null,"number":12893,"object_type":"M","observation_date":"1983 10 08.40478","observatory_code":"413","original_record":"12893J98Q55S   1983 10 08.40478 20 52 03.89 -15 47 20.0                 a3020413"},{"designation":"1998 QS55","discovery_flag":null,"number":12893,"object_type":"M","observation_date":"1983 10 08.44645","observatory_code":"413","origi ...
```
@mommermi - This file needs to be added to the `paths_test` list in `mpc/setup_package.py`, that should fix the current test failure.

mommermi replied:
Yes, good point. I added it to the list. Thanks for pointing this out!

bsipocz commented:
@mommermi - The failures still look relevant, could you have a look please?

mommermi commented:
@mkelley do you mind having a quick look at this?

mkelley commented:
I'm happy to see this added!  Checking it out...  works for asteroid 9530 :+1: 

I'm having trouble with 2P/Encke:
```
In [6]: MPC.get_observations(number=2, comettype='P')
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-a1c83dc880fd> in <module>()
----> 1 MPC.get_observations(number=2, comettype='P')

/disks/data0/astro/Projects/astroquery/astroquery/utils/class_or_instance.py in f(*args, **kwds)
     23         def f(*args, **kwds):
     24             if obj is not None:
---> 25                 return self.fn(obj, *args, **kwds)
     26             else:
     27                 return self.fn(cls, *args, **kwds)

/disks/data0/astro/Projects/astroquery/astroquery/utils/process_asyncs.py in newmethod(self, *args, **kwargs)
     27             if kwargs.get('get_query_payload') or kwargs.get('field_help'):
     28                 return response
---> 29             result = self._parse_result(response, verbose=verbose)
     30             self.table = result
     31             return result

/disks/data0/astro/Projects/astroquery/astroquery/mpc/core.py in _parse_result(self, result, **kwargs)
   1173                 # convert comet designations
   1174                 ident = data['desig'][0]
-> 1175                 yr = str(conf.pkd.find(ident[0]))+ident[1:3]
   1176                 let = ident[3]
   1177                 num = str(conf.pkd.find(ident[4]))+ident[5]

/usr/lib/python3/dist-packages/numpy/ma/core.py in __getitem__(self, indx)
   3218         # mask of being reshaped if it hasn't been set up properly yet
   3219         # So it's easier to stick to the current version
-> 3220         dout = self.data[indx]
   3221         _mask = self._mask
   3222 

IndexError: too many indices for array
```

And Pallas:
```
In [9]: MPC.get_observations(number=2)
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-9-74c1c785b82b> in <module>()
----> 1 MPC.get_observations(number=2)

/disks/data0/astro/Projects/astroquery/astroquery/utils/class_or_instance.py in f(*args, **kwds)
     23         def f(*args, **kwds):
     24             if obj is not None:
---> 25                 return self.fn(obj, *args, **kwds)
     26             else:
     27                 return self.fn(cls, *args, **kwds)

/disks/data0/astro/Projects/astroquery/astroquery/utils/process_asyncs.py in newmethod(self, *args, **kwargs)
     27             if kwargs.get('get_query_payload') or kwargs.get('field_help'):
     28                 return response
---> 29             result = self._parse_result(response, verbose=verbose)
     30             self.table = result
     31             return result

/disks/data0/astro/Projects/astroquery/astroquery/mpc/core.py in _parse_result(self, result, **kwargs)
   1182 
   1183             # convert dates to Julian Dates
-> 1184             dates = [d[:10].replace(' ', '-') for d in data['epoch']]
   1185             times = np.array([float(d[10:]) for d in data['epoch']])
   1186             jds = Time(dates, format='iso').jd+times

UnboundLocalError: local variable 'data' referenced before assignment
```

I'm also jealous that asteroid queries just use one keyword, but comet queries use two.  Can I recommend something like the following instead?
```python
"""
        asteroid : int or string
            Asteroid number or provisional designation.  Packed designations are not permitted...
        comet : int or string
            Periodic comet number or provisional designation.  Packed designations are not permitted...
"""
        obj = (str(comet) if comet is not None else str(asteroid)).strip()
        if obj.isnumeric():
            request_payload['number'] = int(obj)
            if comet is not None:
                # all numbered comets are periodic
                request_payload['object_type'] = 'P'
        else:
            request_payload['designation'] = obj
            if comet is not None:
                # all comet designations start with the orbit type
                request_payload['object_type'] = obj[0]    
```

How do we use this to query a comet fragment?  Compare: https://www.minorplanetcenter.net/db_search/show_object?object_id=51P-A to https://www.minorplanetcenter.net/db_search/show_object?object_id=51P-B

mommermi commented:
Thanks, @mkelley, for this thorough review! I fixed the issues that occurred and implemented your recommendation. Using regex matching, the method is now able to distinguish between an asteroid identifier and a comet identifier. Docs and tests have been updated accordingly.

mkelley commented:
Excellent.

maxmahlke commented:
There is a `print(data)` call in [mpc/core.py](https://github.com/astropy/astroquery/blob/fd6aafe11fd69b83da6ae41d97c2cd5c9bfd0b27/astroquery/mpc/core.py#L1193) which always outputs the returned data when querying for asteroids. This is not intended, right?

mommermi commented:
@maxmahlke Good catch! I thought I got rid of all of those...

bsipocz commented on astroquery/mpc/core.py:
```diff
@@ -1,13 +1,17 @@
 import astropy.units as u
-from astropy.coordinates import EarthLocation, Angle
+from astropy.coordinates import EarthLocation, Angle, SkyCoord
+from astropy._erfa.core import ErfaWarning
```
this usage here prompted me to make the ErfaWarning exposed somewhere directly, so it's not available from astropy.utils.exceptions in 2.0.12 and 3.1.2. Given that we still support older ones here a try except will be needed for a while, but it's preferred to not use anything directly from `_erfa`

mommermi replied:
So you mean something like this:

```
try:
    from astropy.utils.exceptions import ErfaWarning
except ImportError:
    from astropy._erfa.core import ErfaWarning
```
? Or do you want me to get rid of `astropy._erfa` entirely?

bsipocz replied:
sorry, I didn't notice this reply. I'm afraid yes, we need to do the try except as the warning became available in `astropy.utils.exceptions` very recently in 2.0.12 (triggered by this PR). Maybe add a line of comment before the _erfa import that this usage is discouraged and should be removed once we require 2.0.12/3.1.2+ in astroquery.

Alternatively we could go bold and ignore all `AstropyUserWarnings`, though I'm not sure that's a wise choice.

bsipocz commented on astroquery/mpc/core.py:
```diff
@@ -762,6 +769,183 @@ def _args_to_ephemeris_payload(self, **kwargs):
+
+        # identify target type and identifier
+
+        pat = ('(^[0-9]*$)|'  # [0] asteroid number
```
So great for future maintenance that you added comments here, thanks!

bsipocz commented on docs/mpc/mpc.rst:
```diff
@@ -2,23 +2,26 @@
-Minor Planet Center Queries (`astroquery.mpc`)
-**********************************************
+*************************************************************************
+Minor Planet Center Queries (`astroquery.mpc`/astroquery.solarsystem.MPC)
```
maybe linkyfy `astroquery.solarsystem.MPC`, too

mommermi replied:
good point, done!

bsipocz commented:
@mommermi - This looks great to me, ready to go from your side? 

A rebase would be nice to get rid of the merge commits, and addressing my erfa related comment, but I'm happy to do the latter separately.

mommermi commented:
Hmm. That rebase didn't seem to make it better...

bsipocz commented:
It hasn't got rebased but upstream merged in.

mommermi commented:
The build error must be due to network hiccups - @bsipocz could you please restart the CI builds? If those pass, this is ready to be merged.

mommermi commented:
@bsipocz @keflavich could either of you please restart the CI builds? Something looks fishy to me...

keflavich commented:
@mommermi done

mkelley commented:
Hi @mommermi ,

I got an error here, similar to the one at NASA-Planetary-Science/sbpy#130

```python
In [4]: MPC.get_observations('P/2010 WK')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-5c9561965d9d> in <module>()
----> 1 MPC.get_observations('P/2010 WK')

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/utils/class_or_instance.py in f(*args, **kwds)
     23         def f(*args, **kwds):
     24             if obj is not None:
---> 25                 return self.fn(obj, *args, **kwds)
     26             else:
     27                 return self.fn(cls, *args, **kwds)

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/utils/process_asyncs.py in newmethod(self, *args, **kwargs)
     24             verbose = kwargs.pop('verbose', False)
     25 
---> 26             response = getattr(self, async_method_name)(*args, **kwargs)
     27             if kwargs.get('get_query_payload') or kwargs.get('field_help'):
     28                 return response

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/utils/class_or_instance.py in f(*args, **kwds)
     23         def f(*args, **kwds):
     24             if obj is not None:
---> 25                 return self.fn(obj, *args, **kwds)
     26             else:
     27                 return self.fn(cls, *args, **kwds)

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/mpc/core.py in get_observations_async(self, targetid, comettype, get_mpcformat, get_raw_response, cache)
    908         if len(m) == 0:
    909             raise ValueError(('Cannot interpret target '
--> 910                               'identifier "{}".').format(targetid))
    911         else:
    912             m = m[0]

ValueError: Cannot interpret target identifier "P/2010 WK".
```

mkelley commented:
Sorry, @mommermi , I've got another one for you :(

Does c/2013 US10 work on your end?

```In [3]: MPC.get_observations('C/2013 US10')
/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py:1321: StringTruncateWarning: truncated right side string(s) longer than 7 character(s) during assignment
  self.columns[item][:] = value
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-ab00006ff0e5> in <module>()
----> 1 MPC.get_observations('C/2013 US10')

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/utils/class_or_instance.py in f(*args, **kwds)
     23         def f(*args, **kwds):
     24             if obj is not None:
---> 25                 return self.fn(obj, *args, **kwds)
     26             else:
     27                 return self.fn(cls, *args, **kwds)

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/utils/process_asyncs.py in newmethod(self, *args, **kwargs)
     27             if kwargs.get('get_query_payload') or kwargs.get('field_help'):
     28                 return response
---> 29             result = self._parse_result(response, verbose=verbose)
     30             self.table = result
     31             return result

/home/msk/local/lib/python3.6/site-packages/astroquery-0.3.10.dev5266-py3.6.egg/astroquery/mpc/core.py in _parse_result(self, result, **kwargs)
   1239             data['mag'].unit = u.mag
   1240 
-> 1241         return QTable(data)
   1242 
   1243 

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in __init__(self, data, masked, names, dtype, meta, copy, rows, copy_indices, **kwargs)
    409 
    410         # Finally do the real initialization
--> 411         init_func(data, names, dtype, n_cols, copy)
    412 
    413         # Whatever happens above, the masked property should be set to a boolean

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in _init_from_table(self, data, names, dtype, n_cols, copy)
    736         cols = list(table.columns.values())
    737 
--> 738         self._init_from_list(cols, names, dtype, n_cols, copy)
    739 
    740     def _convert_col_for_table(self, col):

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in _init_from_list(self, data, names, dtype, n_cols, copy)
    690             cols.append(col)
    691 
--> 692         self._init_from_cols(cols)
    693 
    694     def _init_from_ndarray(self, data, names, dtype, n_cols, copy):

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in _init_from_cols(self, cols)
    763         # plain Table this is self.ColumnClass, but for instance QTable will
    764         # convert columns with units to a Quantity mixin.
--> 765         newcols = [self._convert_col_for_table(col) for col in cols]
    766         self._make_table_from_cols(self, newcols)
    767 

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in <listcomp>(.0)
    763         # plain Table this is self.ColumnClass, but for instance QTable will
    764         # convert columns with units to a Quantity mixin.
--> 765         newcols = [self._convert_col_for_table(col) for col in cols]
    766         self._make_table_from_cols(self, newcols)
    767 

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py in _convert_col_for_table(self, col)
   2833             # identified in the unit (such as u.mag()).
   2834             q_cls = getattr(col.unit, '_quantity_class', Quantity)
-> 2835             qcol = q_cls(col.data, col.unit, copy=False)
   2836             qcol.info = col.info
   2837             col = qcol

/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/units/quantity.py in __new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    373                  isinstance(value.item(() if value.ndim == 0 else 0),
    374                             numbers.Number))):
--> 375             raise TypeError("The value must be a valid Python or "
    376                             "Numpy numeric type.")
    377 

TypeError: The value must be a valid Python or Numpy numeric type.
```

mkelley commented:
And another: I query for C/2008 FK75, and the designation in the resultant table is C/2008 F7:
```
In [133]: MPC.get_observations('C/2008 FK75')
/home/msk/local/lib/python3.6/site-packages/astropy-3.1.1-py3.6-linux-x86_64.egg/astropy/table/table.py:1321: StringTruncateWarning: truncated right side string(s) longer than 7 character(s) during assignment
  self.columns[item][:] = value
Out[133]: 
<QTable masked=True length=3702>
number comettype  desig  note1 note2     epoch             RA                 DEC           mag   phottype observatory
                                           d              deg                 deg           mag                       
int64     str1     str7   str1  str1    float64         float64             float64       float64   str1       str3   
------ --------- ------- ----- ----- ------------- ------------------ ------------------- ------- -------- -----------
    --         C 2008 F7    --     C 2454556.86399         215.899625 -18.485555555555557    18.9        N         703
    --         C 2008 F7    --     C 2454556.87132 215.89874999999998 -18.485055555555558    19.4        N         703
    --         C 2008 F7    --     C 2454556.87863 215.89812499999996 -18.484750000000002    20.0        N         703
```

The return value from `MPC.query_object` is as expected:
```
In [141]: MPC.query_object(target_type='comet', designation='C/2008 FK75')
Out[141]: 
[{'aphelion_distance': None,
  'argument_of_perihelion': '80.46762',
  'ascending_node': '218.2656364',
  'computer_name': 'MPCW',
  'created_at': '2014-04-24T17:11:03Z',
  'designation': 'C/2008 FK75',
  'earth_moid': None,
 ...
```

mommermi commented:
@mkelley thanks for your bug reports! I sorted them out and added related tests. I also added a keyword argument that overrides the automated target type identification. Just in case...

@bsipocz thank you for your review and sorry for my late reply - I hadn't seen the review until Friday...

Once the CI builds succeed, this should be ready for merging.

bsipocz commented:
Thank you @mommermi. The conflicting changelog seems to be messed up again, but rather than rebasing and fixing it again, I suggest to merge my restructuring #1384 first and then this PR.
It may take me a few days I suppose to get back to that one, hope a bit more wait with this one is OK.

mommermi commented:
@bsipocz Sounds good to me! Could you please have a quick look at my reply to your review comments concerning `astropy._erfa` at some point? Thanks!

mkelley commented:
Hi @mommermi, another issue.  I'm using this to get data for 240P.  The returned data ends in October 2018, but there should be data through May 2019:

Compare:
https://www.minorplanetcenter.net/db_search/show_object?object_id=240P

To:
```
>>> tab = astroquery.mpc.MPC.get_observations('240P')
>>> tab.sort('epoch')
>>> tab[-1]
<Row index=2202 masked=True>
number comettype note1  desig  note2     epoch             RA                DEC          mag   phottype observatory
                                           d              deg                deg          mag                       
int64     str1    str1   str7   str1    float64         float64            float64      float64   str1       str3   
------ --------- ----- ------- ----- ------------- ------------------ ----------------- ------- -------- -----------
   240         P    -- 2002 X2     C 2458404.73591 146.30512499999998 28.75338888888889    17.6        N         J22
>>> from astropy.time import Time
>>> Time(2458404.73591, format='jd').iso
'2018-10-13 05:39:42.624'
```

mommermi commented:
@mkelley This is interesting - it works for me:
```
>>> from astroquery.mpc import MPC
>>> from astropy.time import Time
>>> tab = MPC.get_observations('240P')
>>> tab.sort('epoch')
>>> Time(tab['epoch'][-1], format='jd').iso
'2019-05-05 21:35:15.936'
```
Can you try again? Maybe it was some kind of MPC server hiccups?

mkelley commented:
After remembering to set `cache=False`, I did finally get all the data.  Thanks for checking.

bsipocz commented:
@mommermi - could you please rebase? now that #1384 is merged I expect no more changelog conflicts.

bsipocz commented:
now this get worse by picking up unrelated changes from master. Could you please rebase rather than pull/merge?

mommermi commented:
I'll never get this right ... are you available on slack?

bsipocz commented on CHANGES:
```diff
@@ -170,6 +172,9 @@ skyview
 solarsystem
 ^^^^^^^^^^^
 
+- updates to jplhorizons [#1335, #1478] and mpc [#1350]
```
I'm not sure how to deal with the changelog for the aggregating modules. shall we just move the entries from e.g. ``jplhorizon`` to ``solarsystem.jplhorizon``? I'm not sure. @keflavich?

keflavich replied:
sure, why not?  Or just keep them all under `solarsystem` and say stuff like:
```
solarsystem
^^^^^^^^^^^^
- jplhorizons: updates to blahblah [#1335]
- mpc: updates to foobar [#1350]
```
since it is unlikely these subsections in the changelog will grow out of hand

mommermi replied:
Another way would be to simply remove `solarsystem` here, as it simply serves as a collection of tools in other models. I guess this depends on how you want to organize the modules in the future...

bsipocz replied:
in astropy we have `astropy.io.ascii`, `astropy.io.fits`, etc. listed in alphabetical. For some of the modules here that will make sense, e.g. for `esa.gaia` or `cds.vizier`, but I suppose it doesn't make sense for `solarsystem.mpc` or `observatory_archives.cadc`.

While probably it worth having a `solarsystem` entry when a new module is added to that collection. 

Does it make sense?

mommermi replied:
Ok, so changes to `jplhorizons` and `mpc` go under the corresponding module header and in a short form under `solarsystem`? Or do we use only one of the aforementioned?

bsipocz replied:
no, I think the former is enough, and `solarsystem` only gets an entry when e.g. `imcce` is added. No need to change here anything, I'll clean up the changelog before tagging the release.

mommermi replied:
got it, thanks!

bsipocz commented:
Thanks @mommermi!

mommermi commented:
@bsipocz: my pleasure and thanks for your help!
